### PR TITLE
Add trex Unit tests

### PR DESCRIPTION
**Pull request by RamLavi**
This PR is adding unit tests to check the trex stats parsing.

### CODE REVIEW

**Comment by RamLavi** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/dd8952caddfbcc81c51769089184fc9329201fc3..60d292c778094d4533ce242b77571a6dca85af3c): Rebase

[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/fa38ea326c428a756fa93e6f8baeb1c238f89b2a..943604d671999970082d6767493207e9a09551c0): review fixes

File: pkg/internal/checkup/trex/client_test.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+const (
+	trafficGeneratorPacketsPerSecond = "1m"
+	testDuration                     = time.Second
+	verbosePrintsEnabled             = false
+
+	portIdx = trex.SourcePort
+)
+
+func TestGetPortStatsSuccess(t *testing.T) {
+	t.Run("when GetPortStats succeed", func(t *testing.T) {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:44]: Please consider not using a sub-test if there is only one.

File: pkg/internal/checkup/trex/client_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+		"           \r\nrx_cpu_util. : \x1b[32m0.0\x1b[39m% / 0 pps                               total_pps    : 7.99 Mpps              " +
+		"        \r\nasync_util.  : \x1b[32m0\x1b[39m% / 0 bps                                 drop_rate    : \x1b[32m0 bps\x1b[39m     " +
+		"                     \r\ntotal_cps.   : 0 cps                                      queue_full   : \x1b[32m0 pkts\x1b[39m       " +
+		"                  \r\n\r\n" +
+		"trex>Shutting down RPC client" +
+		"\r\n\r\n" +
+		"[root@dpdk-traffic-gen-jscpt trex]# "
+)
+
+type stubExpecter struct {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:245]:
```suggestion
type expecterStub struct {
```

File: pkg/internal/checkup/trex/client_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+				Oerrors:     15,
+				Opackets:    480000000,
+			},
+		}
+		assert.Equal(t, expected, stats, "GetPortStats returned unexpected result")
+	})
+}
+
+func TestGetPortStatsFailure(t *testing.T) {
+	t.Run("when fails on batch error", func(t *testing.T) {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:72]:
```suggestion
	t.Run("when batch execution fails", func(t *testing.T) {
```

File: pkg/internal/checkup/trex/client_test.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+		expectBatchErr := errors.New("failed to run batch")
+		expecter := &stubExpecter{
+			expectBatchErr: expectBatchErr,
+		}
+
+		c := trex.NewClient(expecter, trafficGeneratorPacketsPerSecond, testDuration, verbosePrintsEnabled)
+		stats, err := c.GetPortStats(portIdx)
+
+		assert.ErrorContains(t, err, expectBatchErr.Error())
+		assert.Empty(t, stats)
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:82]: nit:
```suggestion

		stats, err := c.GetPortStats(portIdx)
		assert.ErrorContains(t, err, expectBatchErr.Error())
		assert.Empty(t, stats)
```

File: pkg/internal/checkup/trex/client_test.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	})
+	t.Run("when fails on timeout", func(t *testing.T) {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:84]:
```suggestion
	t.Run("when batch times out", func(t *testing.T) {
```

File: pkg/internal/checkup/trex/client_test.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+		expectTimeoutErr := errors.New("failed on timeout")
+		expecter := &stubExpecter{
+			timeoutErr: expectTimeoutErr,
+		}
+		c := trex.NewClient(expecter, trafficGeneratorPacketsPerSecond, testDuration, verbosePrintsEnabled)
+		stats, err := c.GetPortStats(portIdx)
+
+		assert.ErrorContains(t, err, expectTimeoutErr.Error())
+		assert.Empty(t, stats)
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:93]: nit:
```suggestion

		stats, err := c.GetPortStats(portIdx)
		assert.ErrorContains(t, err, expectTimeoutErr.Error())
		assert.Empty(t, stats)
```

File: pkg/internal/checkup/trex/client_test.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+				Opackets:    480000000,
+			},
+		}
+		assert.Equal(t, expected, stats, "GetPortStats returned unexpected result")
+	})
+}
+
+func TestGetPortStatsFailure(t *testing.T) {
+	t.Run("when fails on batch error", func(t *testing.T) {
+		expectBatchErr := errors.New("failed to run batch")
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:73]:
```suggestion
		expectedBatchErr := errors.New("failed to run batch")
```

File: pkg/internal/checkup/trex/client_test.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+		}
+
+		c := trex.NewClient(expecter, trafficGeneratorPacketsPerSecond, testDuration, verbosePrintsEnabled)
+		stats, err := c.GetPortStats(portIdx)
+
+		assert.ErrorContains(t, err, expectBatchErr.Error())
+		assert.Empty(t, stats)
+	})
+	t.Run("when fails on timeout", func(t *testing.T) {
+		expectTimeoutErr := errors.New("failed on timeout")
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:85]:
```suggestion
		expectedTimeoutErr := errors.New("failed on timeout")
```

File: pkg/internal/checkup/trex/client_test.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+		c := trex.NewClient(expecter, trafficGeneratorPacketsPerSecond, testDuration, verbosePrintsEnabled)
+		stats, err := c.GetPortStats(portIdx)
+
+		assert.ErrorContains(t, err, expectTimeoutErr.Error())
+		assert.Empty(t, stats)
+	})
+}
+
+func TestGetGlobalStatsSuccess(t *testing.T) {
+	t.Run("when GetPortStats succeed", func(t *testing.T) {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:98]: Please consider not using a sub-test when there is only one.

File: pkg/internal/checkup/trex/client_test.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+		expecter := stubExpecter{}
+		c := trex.NewClient(expecter, trafficGeneratorPacketsPerSecond, testDuration, verbosePrintsEnabled)
+
+		stats, err := c.GetGlobalStats()
+		assert.NoError(t, err, "GetGlobalStats returned an error")
+		expected := trex.GlobalStats{
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:104]: nit:
```suggestion
		assert.NoError(t, err, "GetGlobalStats returned an error")

		expected := trex.GlobalStats{
```

File: pkg/internal/checkup/trex/client_test.go [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+		"\r\nowner      |              \x1b[32mroot\x1b[39m \r\nlink       |                UP \r\nstate      |              \x1b[1mIDLE" +
+		"\x1b[22m \r\nspeed      |           10 Gb/s \r\nCPU util.  |              \x1b[32m0.0\x1b[39m% \r\n--         |                " +
+		"   \r\nTx bps L2  |         1.82 Gbps \r\nTx bps L1  |         2.36 Gbps \r\nTx pps     |         3.35 Mpps \r\nLine Util. |   " +
+		"        \x1b[1m23.56 %\x1b[22m \r\n---        |                   \r\nRx bps     |             0 bps \r\nRx pps     |          " +
+		"   0 pps \r\n----       |                   \r\nopackets   |                 0 \r\nipackets   |                 0 \r\nobytes   " +
+		"  |                 0 \r\nibytes     |                 0 \r\ntx-pkts    |            0 pkts \r\nrx-pkts    |            0 pkts " +
+		"\r\ntx-bytes   |               0 B \r\nrx-bytes   |               0 B \r\n-----      |                   \r\noerrors    |      " +
+		"           \x1b[32m0\x1b[39m \r\nierrors    |                 \x1b[32m0\x1b[39m \r\n\r\ntrex>Shutting down RPC client\r\n\r\n[r" +
+		"oot@dpdk-traffic-gen-jscpt trex]# "
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:184]: Please consider using a raw string, and/or break the lines after `\r\n`

File: pkg/internal/checkup/trex/client_test.go [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+	globalStatsCmd    = "cd /opt/trex && echo \"verbose on;stats -g\" | ./trex-console -q\n"
+	globalStatsOutput = "Using 'python3' as Python interpeter\r\n\r\n\r\n-=TRex Console v3.0=-\r\n\r\nType 'help' or '?' for supported a" +
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client_test.go:187]: Same

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @RamLavi.
WDYT about testing the other exported methods of `trex.Client`?

File: pkg/internal/checkup/trex/client_test.go (hunk 1 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:44]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 3 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:72]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 7 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:73]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 4 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:82]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 5 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:84]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 8 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:85]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 6 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:93]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 9 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:98]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 10 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:104]: DONE

File: pkg/internal/checkup/trex/client_test.go (hunk 11 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:184]: I prefer not to. The part that is relevant is the json part (starts with `Server Response` and ends  with the json end). This part is broken appropriately for better readability.

Regarding the rest - we don'r really care about it so we just make sure the line is not too long and break it every set characters.

File: pkg/internal/checkup/trex/client_test.go (hunk 12 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:187]: Same.

File: pkg/internal/checkup/trex/client_test.go (hunk 2 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client_test.go:245]: DONE

**Review by orelmisan — APPROVED** [review]
Thank you

**Pull request merged**